tools/config/main.c

```diff
@@ -0,0 +1,477 @@
+
+            if (cn_root) {
+                set_key_values(cn_root, repo, &argv[optind+2]);
+                if (filename) {
+                    rc = write_file(cn_root, filename);
+                    if (rc != 0)
+                        fail(ERR_SYSTEM, "failed to write file '%s': %s (%d)", filename, strerror(errno), errno);
+
+                    safe_free(filename);
+                } else
+                    cnfmodule_unparse(mod_ini, stdout, cn_root);
+
+                destroy_cnftree(cn_root);
+            }
+        } else if (strcmp(action, "get") == 0) {
+            if (argcount < 2)
+                fail(ERR_CMDLINE, "expected main or repo name\n");
+
+            repo = argv[optind+1];
+
+            if (argcount < 3)
+                fail(ERR_CMDLINE, "expected one setting\n");
+
+            cn_root = get_repo_root(main_config, repo, NULL);
+
+            if (cn_root) {
+                struct cnfnode *cn_repo = find_child(cn_root, repo);
+                if (cn_repo) {
+                    struct cnfnode *cn_keyval = find_child(cn_repo, argv[optind+2]);
+                    if (cn_keyval)
+                        printf("%s\n", cn_keyval->value);
+                    else
+                        fail(ERR_NO_SETTING, "'%s' not found in '%s'\n", argv[optind+2], repo);
+                } else
+                    fail(ERR_NO_REPO, "repo '%s' not found\n", repo);
+                
+                destroy_cnftree(cn_root);
+            }
+        } else if (strcmp(action, "remove") == 0) {
+            if (argcount < 2)
+                fail(ERR_CMDLINE, "expected main or repo name\n");
+
+            repo = argv[optind+1];
+
+            if (argcount < 3)
+                fail(ERR_CMDLINE, "expected one setting\n");
+
+            cn_root = get_repo_root(main_config, repo, &filename);
+
+            if (cn_root) {
+                remove_keys(cn_root, repo, &argv[optind+2]);
+                if (filename) {
+                    rc = write_file(cn_root, filename);
+                    if (rc != 0)
+                        fail(ERR_SYSTEM, "failed to write file '%s': %s (%d)", filename, strerror(errno), errno);
+
+                    safe_free(filename);
+                } else
+                    cnfmodule_unparse(mod_ini, stdout, cn_root);
+
+                destroy_cnftree(cn_root);
+            }
+        } else if (strcmp(action, "removerepo") == 0) {
+            if (argcount < 2)
+                fail(ERR_CMDLINE, "expected main or repo name\n");
+
+            repo = argv[optind+1];
+
+            cn_root = get_repo_root(main_config, repo, &filename);
+
+            if (cn_root) {
+                remove_repo(cn_root, repo);
+                if (filename) {
+                    if (cn_root->first_child) {
+                        rc = write_file(cn_root, filename);
+                        if (rc != 0)
+                            fail(ERR_SYSTEM, "failed to write file '%s': %s (%d)", filename, strerror(errno), errno);
```

Comment:
can we have all error strings in a macro so printing will be more readable

Reply:
Same here - macros make it harder to read.